community/planning/service_timer.md

```diff
@@ -0,0 +1,244 @@
+mandate that a greater amount of information be stored in the database with the
+ability to re-start should the process be restarted or an internal error occurs.
+
+## Guide-level Explanation
+
+### Implementing service components
+
+Two traits must be implemented by service: `TimerHandler` and `TimerFilter`.
+
+A `TimerHandler` is run when there is work to be performed and together with
+`MessageHandler`, will contain most of the logic of the service. When run, a
+`TimerHandler` is provided a sender and a service ID. For example:
+
+```rust
+pub struct MyTimerHandler { ... }
+
+impl TimerHandler for MyTimerFilter {
+    type Message = MyMessage;
+
+    fn handle_timer(
+        &mut self,
+        sender: &dyn MessageSender<Self::Message>,
+        service: FullyQualifiedServiceId,
+    ) -> Result<(), InternalError> {
+        // determine what to do (possibly from the database) and execute logic here, which can
+        // include sending messages to services on the same circuit
+    }
+}
+```
+
+A `TimerFilter` is run to determine which services have work to perform.
+
+```rust
+pub struct MyTimerFilter { ... }
+
+impl TimerFilter for MyTimerFilter {
+    fn filter(&self) -> Result<Vec<FullyQualifiedServiceId>, InternalError> {
+         // query the database and return a list of service ids which should be woken up
+    }
+}
+```
+
+### Runtime execution and integration within Splinter
+
+A `Timer` handles execution of both `TimerFilter`s and `TimerHander`s. The
+`Timer` wakes up periodically and runs a `TimerFilter` for each service type
+supported (example service types: echo, scabbardv2). Each filter returns a list
+of service IDs and for each service ID returned, the `Timer` executes a
+`TimerHandler` of the corresponding type.
+
+## Reference-level Explanation
+
+### TimerFilter Trait
+The `TimerFilter` is used to determine the list of `FullyQualifiedServiceId`
+that needs to be handled. Each service type will need their own `TimerFilter`.
+
+```rust
+pub trait TimerFilter: Routable{
+    fn filter(&self) -> Result<Vec<FullyQualifiedServiceId>, InternalError>;
+}
+```
+
+The `TimerFilter` must be `Routable` meaning it has an associated service types.
+
+```rust
+pub trait Routable {
+    fn service_types(&self) -> &[ServiceType];
+}
+```
+
+For example, the `EchoService` will return the list of all services that have a
+peer and are in the `Finalized` state. See the
+[EchoService]({% link community/planning/echo_service_design.md%}).
+for more information
+
+### TimerHandler Trait
+The `TimeHandler` is in charge of executing any work that should be done on some
+interval.
+
+```rust
+pub trait TimerHandler {
+    type Message;
+
+    fn handle_timer(
+        &mut self,
+        sender: &dyn MessageSender<Self::Message>,
+        service: FullyQualifiedServiceId,
+    ) -> Result<(), InternalError>;
+
+    fn into_handler<C, R>(
+self,
+converter: C
+) -> IntoTimerHandler<Self, C, Self::Message, R>
```

Comment:
looks like the indenting got a little wonky?

Reply:
fixed